Emby.Server.Implementations/Session/SessionWebSocketListener.cs

```diff
@@ -136,85 +146,121 @@ private void EnsureController(SessionInfo session, IWebSocketConnection connecti
+            _logger.LogDebug("WebSocket {0} is closed.", webSocket);
+            RemoveWebSocket(webSocket);
         }
 
         /// <summary>
         /// Adds a WebSocket to the KeepAlive watchlist.
         /// </summary>
         /// <param name="webSocket">The WebSocket to monitor.</param>
-        private async void KeepAliveWebSocket(IWebSocketConnection webSocket)
+        private void KeepAliveWebSocket(IWebSocketConnection webSocket)
         {
-            _webSockets.TryAdd(webSocket, 0);
-            webSocket.Closed += _webSocket_Closed;
-            webSocket.LastKeepAliveDate = DateTime.UtcNow;
-
-            // Notify WebSocket about timeout
-            try
+            lock (_webSocketsLock)
             {
-                await SendForceKeepAlive(webSocket);
+                if (!_webSockets.Add(webSocket))
+                {
+                    _logger.LogWarning("Multiple attempts to keep alive single WebSocket {0}", webSocket);
+                    return;
+                }
+                webSocket.Closed += OnWebSocketClosed;
+                webSocket.LastKeepAliveDate = DateTime.UtcNow;
+
+                StartKeepAlive();
             }
-            catch (WebSocketException exception)
+
+            // Notify WebSocket about timeout
+            SendForceKeepAlive(webSocket);
+        }
+
+        /// <summary>
+        /// Removes a WebSocket from the KeepAlive watchlist.
+        /// </summary>
+        /// <param name="webSocket">The WebSocket to remove.</param>
+        private void RemoveWebSocket(IWebSocketConnection webSocket)
+        {
+            lock (_webSocketsLock)
             {
-                _logger.LogDebug(exception, "Error sending ForceKeepAlive message to WebSocket.");
+                if (!_webSockets.Remove(webSocket))
+                {
+                    _logger.LogWarning("WebSocket {0} not on watchlist.", webSocket);
+                }
+                else
+                {
+                    webSocket.Closed -= OnWebSocketClosed;
+                }
             }
-
-            StartKeepAliveTimer();
         }
 
         /// <summary>
-        /// Starts the KeepAlive timer.
+        /// Starts the KeepAlive watcher.
         /// </summary>
-        private void StartKeepAliveTimer()
+        private void StartKeepAlive()
         {
-            if (_keepAliveTimer == null)
+            lock (_keepAliveLock)
             {
-                _keepAliveTimer = new Timer(
-                    KeepAliveSockets,
-                    null,
-                    TimeSpan.FromSeconds(WebSocketLostTimeout * TimerFactor),
-                    TimeSpan.FromSeconds(WebSocketLostTimeout * TimerFactor)
-                );
+                if (_keepAliveCancellationToken == null)
+                {
+                    _keepAliveCancellationToken = new CancellationTokenSource();
+                    // Start KeepAlive watcher
+                    var task = RepeatAsyncCallbackEvery(
```

Comment:
why declare a `task` if you're not using it afterwards? doesn't C# allow you to _not_ assign a return value if you don't need it?